### DISCUSSION TITLE
Should we use TEOS-10?

### DISCUSSION THREAD

**Discussion by adele-morrison** [mom-ocean/MOM6#1592] [Q&A]
COSIMA is developing some new MOM6 configs, and we're wondering which equation of state to use. @StephenGriffies tells me that GFDL is planning to move to TEOS10 and that @Hallberg-NOAA is updating the code to address some issues and a bug in the present TEOS10 implementation. @Hallberg-NOAA, any further details on these updates and the status of the TEOS10 implementation in MOM6 would be great. Thanks!

**Comment by Hallberg-NOAA** [comment]
Thank you for this question @adele157.  The MOM6 code on the main branch has 5 different equations of state, and I will soon be adding two more.  So it is perfectly natural to ask (1) why so many equations of state, and (2) which one should be used?

  The linear equation of state is very simple and should only be used for very idealized problems.

  The UNESCO equation of state and the Wright equation of state are both fits to the 1981 UNESCO equation of state as described in the appendix of Gill (1982), for example, and both use potential temperature relative to the surface and practical salinity as their state variables (although we treat practical salinity in the model as though it were absolute salinity).  The "UNESCO" equation of state is actually from Jackett and McDougall (1995), using the same functional form as the original UNESCO publication, but with potential temperature replacing in-situ temperature.  The Wright (JAOT, 1997) equation of state is also a fit to the UNESCO equation of state, but uses a more convenient functional form that can be analytically integrated, making it computationally efficient.

  There are actually 2 forms of the Wright equation of state, one with a fit over the full range of validity of UNESCO equation of state and another using the reduced range of pressure from 0 to 5000 dbar, temperature from -2 to 30 degC, and salinity from 28 to 38 PSU.  What is coded up in the main version of MOM6 is this reduced range fit.  We are adding two more equations of state based on the Wright equation of state - WRIGHT_FULL and WRIGHT_RED - which use the two fits from Wright (1997), but also add parentheses to specify the order of arithmetic across compliers and levels of optimization, but are retaining the original WRIGHT equation of state (equivalent to WRIGHT_RED) to be able to reproduce existing solutions.

  This brings us the the TEOS10 and "NEMO" equations of state, which use conservative temperature and absolute salinity as their state variables.  TEOS10 is based on the common package, but it is very computationally expensive.  The "NEMO" equation of state is actually from Roquet et el. (Ocean Modelling, 2015) and it is a fit over the same range of validity as TEOS10, but uses a simple polynomial form that is more efficient to evaluate.  The one caveat with both of these equations of state is that because they use different state variables, we need to be careful about doing the conversions when initializing from or comparing with oceanic data, and we need to use a different functional form of the freezing point calculation that works in the same state variables.  Between the two options, I would encourage you to consider the NEMO equation of state for efficiency, but with a two caveats.

  1. The density derivatives code currently in MOM_EOS_NEMO are mildly inconsistent with the actual derivatives of the density code because they specify separate sets of hard-coded coefficients without enough precision.  I have an updated version of the NEMO / Roquet equation of state that corrects this and also adds parentheses to specify the order of arithmetic.  This is working and is passing our self-consistency testing, but I am waiting for a large PR with lots of documentation from dev/gfdl to main to be handled before I want to add this answer-changing update.  To the best of our knowledge, no one is using the NEMO equation of state yet for active production runs, so we are not planning to keep the existing version around for backward compatibility.
  2.  Roquet et al. kindly provide two fits to the TEOS10 equation of state - one is a polynomial for density, while the other is a polynomial for specific volume.   The former is what is coded up in MOM_EOS_NEMO, but if you have a non-Boussinesq model, the latter is is the one that will be more efficient.  Code to implement the latter is still a work in progress, but it might become available via the main branch of MOM6 with the same timeframe as the updated version of MOM_EOS_NEMO.

  I realize that this is a long answer to a short question, but hopefully you find this informative.  A shorter version of the answers might be "if you can wait a few weeks for some pending code fixes, I would suggest you try the updated NEMO (Boussinesq) or new ROUQUET_SPV (non-Boussinesq) equation of state, rather than TEOS10."

**Comment by adele-morrison** [comment]
Thanks Bob - both for this detailed response and for your work improving the EOS options! We will stick with WRIGHT for our COSIMA new config testing now, and will aim to switch to NEMO when the pending code fixes are merged.

**Comment by aekiss** [comment]
Bob's EOS updates are in this PR: https://github.com/NOAA-GFDL/MOM6/pull/331

**Comment by Hallberg-NOAA** [comment]
At this point what we have on my PR: https://github.com/NOAA-GFDL/MOM6/pull/331 is a fully functioning first draft.  There are some mathematically equivalent changes that we still want to make for computational efficiency, clarity and robustness, including revising some of the repeated rescaling of variables in the Roquet_rho (a.k.a. NEMO) and Roquet_SpV equations of state, and also adapting a version of the conservative-temperature to potential-temperature conversion functions to MOM6 so that we can add parentheses for numerical accuracy and reproducibility.  These will be mathematically equivalent, but will change the order of sums (and hence answers) in some cases.  You can start using this code now, but please be advised that there may still be some minor answer changes before this gets merged into the dev/gfdl and then the main branches of MOM6.

**Comment by adele-morrison** [comment]
Thanks Bob! We'll give it a go.
